### ISSUE TITLE
Standardize input dimensions for DEMs

### ISSUE THREAD

**Issue by erikmannerfelt** [GlacioHack/xdem#89]
Should `xdem` allow DEMs with the shape `(1, height, width)`, `(height, width)` or either of the two? Right now, there's no real consensus on which one works (or at least the first might not always work, but will in some cases).

1.  If we only want to allow `(height, width)`, all functions should have:
```python
assert len(dem.shape) == 2, f"Given DEM has shape '{dem.shape}'. Expected 2D array"
```

2. If we want to allow both, but it should conform to `(height, width)`:
```python
demc = dem.squeeze()  # Create a squeezed copy and work on the copy

assert len(dem.shape) == 2,  f"Given DEM has shape '{dem.shape}'. Expected 2D array or 3D array with shape[0] == 1"
```

3. Finally, if we want both but it should conform  to `(1, height, width)`:
```python
if len(dem.shape) == 3 and dem.shape[0] == 1:
    demc = dem.copy()
elif len(dem.shape) == 2:
    demc = dem.reshape((1,) + dem.shape)
else:
    raise AssertionError(f"Given DEM has shape '{dem.shape}'. Expected 2D array or 3D array with shape[0] == 1")
```

I vote for either 1 or 2. What do you guys think?

It could be implemented in a `def _check_dem(dem) -> None:` to avoid repeating code everywhere.

**Comment by adehecq** [comment]
Yes, I think we should force one of the cases. I would favor using only (height, width) as it makes many steps easier...
But what create a copy each time?
And yes, the check should be written as part of a separate function, that would be called at the beginning of each function.

**Comment by erikmannerfelt** [comment]
What about squeezing automatically then for GeoUtils compatibility? So (1, height, width) automatically becomes (height, width).

Also, a dtype check could maybe be done: make sure it's a float32 or float64 raster, and if not, convert it appropriately.

Yeah I guess we should minimize copying, but I just want to make sure no edits are done on the input... Why can't python have immutable borrows!?!?!

**Comment by adehecq** [comment]
Maybe we should just re-use [get_array_and_mask](https://github.com/GlacioHack/xdem/blob/8faaef1185dcd910a65183def5c4d2a10d8c280e/xdem/spatial_tools.py#L25) and add a step to force a dimension of 2? Then we would simply call this function each time we want to enforce the shape/dtype.
The function already returns a copy if the dtype needs to be changed, but otherwise not.
